### PR TITLE
Add visualization of ASTs to CLI & REPL 

### PR DESCRIPTION
**Pull request by jpschorr**
Add visualization of ASTs to CLI & REPL 
- Make modules for existing functionality of `error` and `repl`
- Add actual arg parsing to CLI
- Make REPL a sub-mode initiated by a CLI command
- Add partial (lots of `todo!()`s) translation of AST into graphviz
- Add FFI to graphviz to layout & render
- Add `display` mode which prints image to console
---
*NOTE*: You will need to `brew install graphviz` or system-equivalent prior to the following.
---
 For example, running:

```cargo run --release --features visualize -- ast -Tpng  "SELECT num.*, num BETWEEN 2 AND 15 FROM (SELECT tbl.a[8] + ((4/-2)+(7*sin(7))-2.0) - xform(1,position('x' in 'yxz')), NULL FROM tbl WHERE tbl.b LIKE 'foo%') as num" >  /tmp/ast.png```

 shows
ast

Or, run 

```cargo run --release --features visualize -- ast -Tdisplay  "SELECT num.*, num BETWEEN 2 AND 15 FROM (SELECT tbl.a[8] + ((4/-2)+(7*sin(7))-2.0) - xform(1,position('x' in 'yxz')), NULL FROM tbl WHERE tbl.b LIKE 'foo%') as num"```

 to display the image directly in the terminal (for any terminal supported by [viuer](https://crates.io/crates/viuer/0.5.3))

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-cli/Cargo.toml [hunk 1]
```diff
@@ -41,14 +41,45 @@ supports-unicode = "1.0.2"
...
+clap = { version = "3.2.13", features = ["derive"] }
+
+# serde
+serde = { version ="1.*", features = ["derive"], optional = true }
+serde_json = { version ="1.*", optional = true }
 
+### Dependencies for the `render` feature
+viuer = { version ="0.6.1", features = ["sixel"], optional = true }
+image = { version ="0.24.2", optional = true }
+graphviz-sys = { version ="0.1.3", optional = true }
```
Comment (alancai98) [inline partiql-cli/Cargo.toml:57]: FYI, if you're running this on macOS and run into the following error:

```
❯ cargo run --release --features visualize -- ast -Tdisplay "SELECT num.*, num BETWEEN 2 AND 15 FROM (SELECT tbl.a[8] + ((4/-2)+(7*sin(7))-2.0) - xform(1,position('x' in 'yxz')), NULL FROM tbl WHERE tbl.b LIKE 'foo%') as num"
   Compiling graphviz-sys v0.1.3
   Compiling resvg v0.23.0
   Compiling partiql-parser v0.0.0 (/Users/caialan/wksp/partiql-lang-rust/partiql-parser)
   Compiling viuer v0.6.1
   Compiling termbg v0.4.1
The following warnings were emitted during compilation:

warning: couldn't execute `llvm-config --prefix` (error: No such file or directory (os error 2))
warning: set the LLVM_CONFIG_PATH environment variable to the full path to a valid `llvm-config` executable (including the executable itself)

error: failed to run custom build command for `graphviz-sys v0.1.3`

Caused by:
  process didn't exit successfully: `/Users/caialan/wksp/partiql-lang-rust/partiql-cli/target/release/build/graphviz-sys-241bfa87d370ff4f/build-script-build` (exit status: 101)
  --- stdout
  cargo:rustc-link-lib=gvc
  cargo:rustc-link-lib=cgraph
  cargo:rustc-link-lib=pathplan
  cargo:rustc-link-lib=cdt
  cargo:warning=couldn't execute `llvm-config --prefix` (error: No such file or directory (os error 2))
  cargo:warning=set the LLVM_CONFIG_PATH environment variable to the full path to a valid `llvm-config` executable (including the executable itself)

  --- stderr
  ./wrapper.h:1:10: fatal error: 'graphviz/gvc.h' file not found
  ./wrapper.h:1:10: fatal error: 'graphviz/gvc.h' file not found, err: true
  thread 'main' panicked at 'Unable to generate bindings: ()', /Users/caialan/.cargo/registry/src/github.com-1ecc6299db9ec823/graphviz-sys-0.1.3/build.rs:43:14
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
```

You may also need to install `graphviz` (i.e. `brew install graphviz`). Credit to Josh for the fix.

File: partiql-cli/src/visualize/ast_to_dot.rs [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+        let id = out.node_auto_labelled(lbl).id();
+
+        self.to_dot(out, &ast.func_name).edges(out, &id, "name");
+        self.to_dot(out, &ast.args).edges(out, &id, "args");
+
+        vec![id]
+    }
+}
+
+impl ToDot<ast::Path> for AstToDot {
```
Comment (alancai98) [inline partiql-cli/src/visualize/ast_to_dot.rs:554]: I played around a bit with some of the conformance tests and noticed the ast for some of the expressions may not be correct. Using the "[complicated nested path](https://github.com/partiql/partiql-tests/blob/ef3d21be1250423d321299bb276455a10b187a21/partiql-tests-data/success/syntax/primitives/path-expression.ion#L57-L65)" test from `partiql-tests`, I got the following result:

```
❯ cargo run --release --features visualize -- ast -Tpng '''a.b.c['item']."d"[5].e['s'].f[1+2]''' > /tmp/ast.png
```
ast

The parsed AST output by `parse_partiql` doesn't include nodes along the path other than the start and final nodes (for this example test, only has `a`, and `f[1+2]`).

**Review by alancai98 — COMMENTED** [review]
Very cool work here. The visualization makes the AST (and future IR) much more comprehensible and easier to view the parser's output. Left mostly minor comments and a possible parser bug (related to path expressions).

File: partiql-cli/Cargo.toml (hunk 1 above)
Comment (alancai98) [reply partiql-cli/Cargo.toml:57]: Could add the `brew install graphviz` as a step in the README to run locally. Also may need to test on other OS

File: partiql-cli/src/visualize/ast_to_dot.rs (hunk 2 above)
Comment (am357) [reply partiql-cli/src/visualize/ast_to_dot.rs:554]: Seems to be a bug; cut #159 for looking into this.

File: partiql-cli/src/repl.rs [hunk 3]
```diff
@@ -106,91 +107,27 @@ impl Highlighter for PartiqlHelper {
...
-            _ => {
-                todo!("Not yet handled {:?}", err);
-            }
-        }
-    }
-}
-
 impl Validator for PartiqlHelper {
     fn validate(&self, ctx: &mut ValidationContext) -> rustyline::Result<ValidationResult> {
+        // TODO remove this command parsing hack do something better
```
[the remark this answers is not in the document]
Comment (jpschorr) [reply partiql-cli/src/repl.rs:112]: https://github.com/partiql/partiql-lang-rust/issues/160

File: partiql-cli/src/visualize/render.rs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    let c_graph_str = std::ffi::CString::new(graph_str).expect("cstring new failed");
+    let c_dot = std::ffi::CString::new("dot").expect("cstring new failed");
+    let c_format = std::ffi::CString::new(format.as_ref()).expect("cstring new failed");
+
+    unsafe {
+        let gvc = gv::gvContext();
+        // TODO gvParseArgs to pass 'theme' colors, etc?
+        //    See section 4 of https://www.graphviz.org/pdf/libguide.pdf
+        //    See `dot --help`
+        let g = gv::agmemread(c_graph_str.as_ptr());
```
[the remark this answers is not in the document]
Comment (jpschorr) [reply partiql-cli/src/visualize/render.rs:48]: #161

**Review by alancai98 — APPROVED** [review]

**Pull request merged**